payjoin-cli/src/app/v2/mod.rs

```diff
@@ -248,64 +237,183 @@ impl App {
+        let mut interrupt = self.interrupt.clone();
+        let receiver = tokio::select! {
+            res = self.long_poll_fallback(session, persister) => res,
+            _ = interrupt.changed() => {
+                println!("Interrupted. Call the `resume` command to resume all sessions.");
+                return Err(anyhow!("Interrupted"));
+            }
+        }?;
+        Ok(receiver)
+    }
+
+    async fn check_proposal(
         &self,
         proposal: Receiver<UncheckedProposal>,
-    ) -> Result<Receiver<PayjoinProposal>, Error> {
+        persister: &ReceiverPersister,
+    ) -> Result<()> {
         let wallet = self.wallet();
+        let proposal = proposal
+            .check_broadcast_suitability(None, |tx| Ok(wallet.can_broadcast(tx)?))
+            .save(persister)?;
 
-        // in a payment processor where the sender could go offline, this is where you schedule to broadcast the original_tx
-        let _to_broadcast_in_failure_case = proposal.extract_tx_to_schedule_broadcast();
+        self.check_inputs_not_owned(proposal, persister).await
+    }
 
-        // Receive Check 1: Can Broadcast
+    async fn check_inputs_not_owned(
+        &self,
+        proposal: Receiver<MaybeInputsOwned>,
+        persister: &ReceiverPersister,
+    ) -> Result<()> {
+        let wallet = self.wallet();
         let proposal =
-            proposal.check_broadcast_suitability(None, |tx| Ok(wallet.can_broadcast(tx)?))?;
-        log::trace!("check1");
+            proposal.check_inputs_not_owned(|input| Ok(wallet.is_mine(input)?)).save(persister)?;
+        self.check_no_inputs_seen_before(proposal, persister).await
+    }
+
+    async fn check_no_inputs_seen_before(
+        &self,
+        proposal: Receiver<MaybeInputsSeen>,
+        persister: &ReceiverPersister,
+    ) -> Result<()> {
+        let proposal = proposal
+            .check_no_inputs_seen_before(|input| Ok(self.db.insert_input_seen_before(*input)?))
+            .save(persister)?;
+        self.identify_receiver_outputs(proposal, persister).await
+    }
+
+    async fn identify_receiver_outputs(
+        &self,
+        proposal: Receiver<OutputsUnknown>,
+        persister: &ReceiverPersister,
+    ) -> Result<()> {
+        let wallet = self.wallet();
+        let proposal = proposal
+            .identify_receiver_outputs(|output_script| Ok(wallet.is_mine(output_script)?))
+            .save(persister)?;
+        self.commit_outputs(proposal, persister).await
+    }
 
-        // Receive Check 2: receiver can't sign for proposal inputs
-        let proposal = proposal.check_inputs_not_owned(|input| Ok(wallet.is_mine(input)?))?;
-        log::trace!("check2");
+    async fn commit_outputs(
+        &self,
+        proposal: Receiver<WantsOutputs>,
+        persister: &ReceiverPersister,
+    ) -> Result<()> {
+        let proposal = proposal.commit_outputs().save(persister)?;
+        self.contribute_inputs(proposal, persister).await
+    }
```

Comment:
I don't find way these are chained to be intuitive. Why does `commit_outputs` call `contribute_inputs`? `contribute_inputs` doesn't seem like a part of `commit_outputs` at all.

and so on for each of these transitions. Because of this, I don't find thier separation into functions to be improving readability and would prefer them to be called imperatively in a list.

Reply:
Perhaps we just need a naming change. e.g `process_commit_outputs_state`.
Alternatively what can be done is to return the sumtype and iterate in `process_receiver_proposal` until the session fails or succeeds. I find this to be less intuitive.

Reply:
I don't think a rename to `process_commit_output_state` resolves my concern here.

would a single function that looped over a match on the sumtype, where each branch called a process function work?  The weird part is the nested chain even more than the fact that the function calls another function unrelated to the parent's name.